### PR TITLE
tables: adding user_groups table for Windows

### PR DESCRIPTION
**Pull request by muffins**
This ports the `user_groups` table to Windows. Big shout-out to @dunter for supplying the original iteration of this code base, thanks!

### CODE REVIEW

**Comment by osqueryer** [comment]
:no_good: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [3247](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3247/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [6712](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6712/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [6713](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6713/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [3598](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3598/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [3601](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3601/)) failed one or more tests (**Windows**).

**Comment by muffins** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [3248](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3248/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [3602](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3602/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [3604](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3604/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [6715](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6715/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [3607](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3607/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [6717](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6717/)) failed one or more tests (**Linux**).

**Comment by muffins** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [3608](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3608/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [3609](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3609/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [6719](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6719/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [1201](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1201/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [1202](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1202/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 181d1becaaf46e381a4c7884192da0e4a03faab8 (Job results: [6720](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6720/)) failed one or more tests (**Linux**).

**Comment by muffins** [comment]
ok to test

**Comment by muffins** [comment]
This addresses #3668

**Comment by osqueryer** [comment]
:no_good: The commit 74a79d681cc482bea87e3e47ea3c55e75aeb346e (Job results: [3322](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3322/)) failed the code audit or documentation test.

File: osquery/tables/system/windows/user_groups.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                                1,
+                                (LPBYTE*)&ginfo,
+                                MAX_PREFERRED_LENGTH,
+                                &numGroups,
+                                &dwTotalUserGroupMembers);
+
+    if ((ret != NERR_Success && ret != ERROR_MORE_DATA) || ginfo == nullptr) {
+      VLOG(1) << " NetUserGetLocalGroups failed for user " << user << " with "
+              << ret;
+      continue;
```
Comment (fmanco) [inline osquery/tables/system/windows/user_groups.cpp]: Shouldn't this be a break?

File: osquery/core/windows/process_ops.cpp [hunk 2]
```diff
@@ -109,6 +109,50 @@ int getGidFromSid(PSID sid) {
...
+    LOG(INFO) << "No account name provided.";
+    return nullptr;
+  }
+
+  // Call LookupAccountNameW() once to retrieve the necessary buffer sizes for
+  // the SID (in bytes) and the domain name (in TCHARS):
+  unsigned long sidBufferSize = 0;
+  unsigned long domainNameSize = 0;
+  auto eSidType = SidTypeUnknown;
+  LookupAccountNameW(nullptr,
```
[the remark this answers is not in the document]
Comment (muffins) [reply osquery/core/windows/process_ops.cpp]: We don't care much about it on this call as we're just getting the required buffer size. I checked the MSDN, and technically this function could be used to derive a SID from most any type of account on Windows, username, groupname, domain name, computername, ... That being said I'd say at most we should maybe change the name of the function, but this wont hurt anything regardless of what `eSidType` ends up being.

File: osquery/tables/system/windows/user_groups.cpp (hunk 1 above)
Comment (muffins) [reply osquery/tables/system/windows/user_groups.cpp]: Wow, this whole `do/while` loop should be removed actually. I had a different thought about what `NetUserGetLocalGroups` was doing. Thanks!

**Comment by osqueryer** [comment]
:no_good: The commit c7382f1f5c5a2eaae85efcd9c4b7f0669d4bfddf (Job results: [3570](https://jenkins.osquery.io/job/osqueryPullRequestAudit/3570/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit b5a61857dc083f273296f80b156f89c845819aed (Job results: [4005](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4005/)) failed one or more tests (**Windows**).

**Pull request merged**